File: examples/event/coremodel-event.c
/*
 * CoreModel EventBus Example
 *
 * Copyright (c) 2026 Corellium Inc.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>

#include <coremodel.h>

static void test_event_update(void *priv, uint64_t data0, uint64_t data1, unsigned initial)
{
    printf("event[%s]: %016lx %016lx%s\n", (char *)priv, data0, data1, initial ? " *" : "");
    fflush(stdout);
}

static const coremodel_event_func_t test_event_func = {
    .update = test_event_update };

int main(int argc, char *argv[])
{
    int res, num, idx;
    void *handle;
    void *cm;
    char **names, *datastr;
    uint64_t data[2];

    if(argc < 2) {
        printf("usage: coremodel-event <address[:port]> <eventname0> [...]\n");
        return 1;
    }

    num = argc - 2;
    names = calloc(sizeof(char *), num);
    if(!names)
        return 1;
    for(idx=0; idx<num; idx++) {
        names[idx] = strdup(argv[2 + idx]);
        if(!names[idx])
            return 1;
    }

    res = coremodel_connect(&cm, argv[1]);
    if(res) {
        fprintf(stderr, "error: failed to connect: %s.\n", strerror(-res));
        return 1;
    }

    for(idx=0; idx<num; idx++) {
        datastr = strchr(names[idx], '=');
        if(datastr)
            *(datastr ++) = 0;

        handle = coremodel_attach_event_name(cm, names[idx], &test_event_func, (void *)names[idx]);
        if(!handle) {
            fprintf(stderr, "error: failed to attach event %s.\n", names[idx]);
            coremodel_disconnect(cm);
            return 1;
        }

        if(datastr) {
            data[0] = strtoul(datastr, &datastr, 0);
            if(*datastr == ',')
                data[1] = strtoul(datastr + 1, NULL, 0);
            else
                data[1] = 0;
            coremodel_event_signal(handle, data[0], data[1], 0);
        }
    }

    coremodel_mainloop(cm, -1);

    coremodel_detach(handle);
    coremodel_disconnect(cm);

    for(idx=0; idx<num; idx++)
        free(names[idx]);
    free(names);

    return 0;
}

File: examples/event/coremodel-event.py
from coremodel import *
import time

from coremodel import CoreModelEvent


class Event(CoreModelEvent):
    def __init__(self,eventname):
        super().__init__(eventname=eventname)

    def update(self, data0, data1, initial):
        print("event[{}] data0 {} data1 {} initial {}".format(self.name, data0, data1, initial))


if __name__ == "__main__":
    if len(sys.argv) < 4:
        print("usage: python3 coremodel-event.py <address> <port> <eventname0|eventname0=data0,data1> ... <eventnameN>")
        sys.exit(1)

    address =  sys.argv[1]
    port = sys.argv[2]
    cm0 = CoreModel("cm0", address, port, "./../../")

    event_list = []
    event_name = []
    event_data0 = []
    event_data1 = []

    for event in sys.argv[3:]:
        idx = event.find('=')
        if idx > 0:
            event_name.append(event[:idx])
            d1 = event[idx+1:].find(',')
            if d1 > 0:
                event_data0.append(int(event[idx + 1:idx + 1 +d1]))
                event_data1.append(int(event[idx + d1 + 2:]))
            else:
                event_data0.append(int(event[idx + 1:]))
                event_data1.append(0)
        else:
            event_name.append(event)
            event_data0.append(0)
            event_data1.append(0)

    for i in range(len(event_name)):
        event_list.append(Event(event_name[i]))

    for event in event_list:
        try:
            cm0.attach(event)
        except NameError as e:
            print(str(e), event.name)
            sys.exit(1)

    if cm0.attached_objs is None:
        sys.exit(1)

    cm0.cycle_time = 1000000

    cm0.start()
    for i in range(10):
        for j in range(len(event_list)):
            try:
                event_list[j].event_signal(event_data0[j], event_data1[j], 0)
            except Exception as e:
                print(str(e), event_list[j].name)
            event_data0[j] = int(not event_data0[j])
            event_data1[j] = int(not event_data1[j])
        time.sleep(1)
    cm0.stop_event.set()
    cm0.join()